compiler/src/dmd/timetrace.d

```diff
@@ -0,0 +1,482 @@
+*/
+module dmd.timetrace;
+
+import dmd.errors;
+import dmd.globals;
+import dmd.location;
+import dmd.root.array;
+import dmd.root.file;
+import dmd.common.outbuffer;
+import dmd.root.string : toDString;
+
+// Thread local profiler instance (multithread currently not supported because compiler is single-threaded)
+TimeTraceProfiler* timeTraceProfiler = null;
+
+// processName pointer is captured
+extern(C++)
+void initializeTimeTrace(uint timeGranularity, uint memoryGranularity, const(char)* processName)
+{
+    assert(timeTraceProfiler is null, "Double initialization of timeTraceProfiler");
+    timeTraceProfiler = new TimeTraceProfiler(timeGranularity, memoryGranularity, processName);
+}
+
+extern(C++)
+void deinitializeTimeTrace()
+{
+    if (timeTraceProfilerEnabled())
+    {
+        object.destroy(timeTraceProfiler);
+        timeTraceProfiler = null;
+    }
+}
+
+pragma(inline, true)
+extern(C++)
+bool timeTraceProfilerEnabled()
+{
+    version (LDC)
+    {
+        import ldc.intrinsics: llvm_expect;
+        return llvm_expect(timeTraceProfiler !is null, false);
+    }
+    else
+    {
+        return timeTraceProfiler !is null;
+    }
+}
+
+const(char)[] getTimeTraceProfileFilename(const(char)* filename_cstr)
+{
+    const(char)[] filename;
+    if (filename_cstr)
+    {
+        filename = filename_cstr.toDString();
+    }
+    if (filename.length == 0)
+    {
+        if (global.params.objfiles.length)
+        {
+            filename = global.params.objfiles[0].toDString() ~ ".time-trace";
+        }
+        else
+        {
+            filename = "out.time-trace";
+        }
+    }
+    return filename;
+}
+
+extern(C++)
+void writeTimeTraceProfile(const(char)* filename_cstr)
+{
+    if (!timeTraceProfiler)
+        return;
+
+    const filename = getTimeTraceProfileFilename(filename_cstr);
+
+    OutBuffer buf;
+    timeTraceProfiler.writeToBuffer(&buf);
+    if (filename == "-")
+    {
+        // Write to stdout
+        import core.stdc.stdio : fwrite, stdout;
+        size_t n = fwrite(buf[].ptr, 1, buf.length, stdout);
+        if (n != buf.length)
+        {
+            error(Loc.initial, "Error writing --ftime-trace profile to stdout");
+            fatal();
+        }
+    }
+    else if (!File.write(filename, buf[]))
+    {
+        error(Loc.initial, "Error writing --ftime-trace profile: could not open '%*.s'", cast(int) filename.length, filename.ptr);
+        fatal();
+    }
+}
+
+// Pointers should not be stored, string copies must be made.
+extern(C++)
+void timeTraceProfilerBegin(const(char)* name_ptr, const(char)* detail_ptr, Loc loc)
+{
+    import dmd.root.rmem : xarraydup;
+    import core.stdc.string : strdup;
+
+    assert(timeTraceProfiler);
+
+    timeTraceProfiler.beginScope(xarraydup(name_ptr.toDString()),
+                                 xarraydup(detail_ptr.toDString()), loc);
+}
+
+extern(C++)
+void timeTraceProfilerEnd()
```

Comment:
These are defined but never used.  Though actually using these would be preferable to a D struct, as it at least would be overridable to allow alternative time tracking engines to be used.

Reply:
As mentioned in the PR description, it's currently closely matching LDC's code. I have lots of ideas to refactor the code to be better DMD style, but I'd like to know from LDC maintainers if they plan on keeping their own copy of timetrace.d or whether they want to share with upstream. In that case we need to keep a C++ interface.

Reply:
> In that case we need to keep a C++ interface.

That is what the second sentence is advocating for. The frontend semantic passes are using a D interface and RAII. It would integrate better if it used these C++ hooks instead.

(It would be even better if each stage being traced had an enum value, rather than passing a string name and detail, but one request at a time to nudge it in the right direction).

Reply:
In LDC, we call these C++ functions in a RAII class and only the RAII class is ever used. It is very bug prone to _not_ use RAII, so the D code also uses a RAII struct.

Reply:
I added an enum value for each stage, and now use D functions (because they take a `delegate`) with help of `scope (exit)` for safety, while also keeping the C++ interface for LDC.